virttest/utils_backup.py

```diff
@@ -0,0 +1,297 @@
+            qemu_io_cmd %= (entry["start"], entry["length"], target_path)
+            process.run(qemu_io_cmd, shell=True)
+    # remove the backing file info for the target image
+    rebase_cmd = "qemu-img rebase -u -f qcow2 -b '' %s" % target_path
+    process.run(rebase_cmd, shell=True)
+
+
+def pull_full_backup_to_file(nbd_export, target_path):
+    """
+    Dump pull-mode full backup data to a file
+
+    :param nbd_export: The nbd export info of the backup job
+    :param target_path: The path of the file to dump the data
+    """
+    cmd = "qemu-img convert -f raw %s -O qcow2 %s" % (nbd_export,
+                                                      target_path)
+    logging.debug("Execute command %s: %s", cmd,
+                  process.run(cmd, shell=True).stdout_text)
+
+
+def get_img_data_size(img_path, driver='qcow2'):
+    """
+    Get the actual data size of a image
+
+    :param img_path: The path to the image
+    :return: The actual data size of the image in 'MB'.
+    """
+    cmd = "qemu-img map -f %s %s --output=json -U" % (driver, img_path)
+    result = process.run(cmd, shell=True).stdout_text
+    json_data = json.loads(result)
+    summary = 0
+    for entry in json_data:
+        if entry["data"]:
+            summary += entry["length"]
+    return summary/1024/1024
+
+
+def get_img_data_map(img_path, driver='qcow2'):
+    """
+    Get the data map of a image
+
+    :prama img_path: The path to the image
+    :return: The data map of the image
+    """
+    cmd = "qemu-img map -f %s %s --output=json -U" % (driver, img_path)
+    result = process.run(cmd, shell=True).stdout_text
+    json_data = json.loads(result)
+    data_map = []
+    for entry in json_data:
+        if entry["data"]:
+            data_map.append(entry)
+    return data_map
+
+
+def dd_data_to_vm_disk(vm, disk, bs, seek, count):
+    """
+    Generate some random data to a vm disk
+
+    :param vm: The vm we'll use
+    :param disk: The disk of the vm we'll use
+    :param bs: The 'bs' param of the 'dd' command
+    :param seek: The 'seek' param of the 'dd' command
+    :param count: The 'count' param of the 'dd' command
+    """
+    session = vm.wait_for_login()
+    dd_cmd = "dd if=/dev/urandom of=%s bs=%s seek=%s count=%s; sync"
+    dd_cmd %= (disk, bs, seek, count)
+    output = session.cmd_output(dd_cmd).strip()
+    logging.debug("Using dd to generate data to %s: %s", disk, output)
+    session.close()
+
+
+def dump_data_to_file(data_map, source_file_path, target_file_path, driver='qcow2'):
+    """
+    Dump real data from the source file to a target file according to data map
+
+    :param data_map: The data map used to indicate which part of the image
+    containing real data
+    :param source_file_path: The path to the source file
+    :param target_file_path: The path to the target file
+    """
+    with open(target_file_path, 'w'):
+        pass
+    cmd = "qemu-io -f %s -c \"read -vC %s %s\" %s -r -U  | sed '$d'"
```

Comment:
should all below logic be put under with open block?

Reply:
yes, updated, thx